contracts/proxy/src/contract.rs

```diff
@@ -335,76 +336,111 @@ pub fn execute_update_guardians(
+                    code_id: instantiation_code_id,
+                    msg: to_binary(&multisig_instantiate_msg)?,
+                    funds: multisig_settings.multisig_initial_funds,
+                    label: "Wallet-Multisig".into(),
+                };
+                let msg = SubMsg::reply_always(instantiate_msg, MULTISIG_INSTANTIATE_ID);
+
+                PENDING_GUARDIAN_ROTATION.remove(deps.storage);
+
+                Ok(Response::new()
+                    .add_submessage(msg)
+                    .add_attribute("action", "Updated wallet guardians: Multisig"))
+            } else {
+                MULTISIG_ADDRESS.save(deps.storage, &None)?;
+                PENDING_GUARDIAN_ROTATION.remove(deps.storage);
+                Ok(Response::new()
+                    .add_attribute("action", "Updated wallet guardians: Non-Multisig"))
             }
-        };
-        MULTISIG_CODE_ID.save(deps.storage, &instantiation_code_id)?;
-        let multisig_instantiate_msg = FixedMultisigInstantiateMsg {
-            voters: addresses_to_voters(&guardians.addresses),
-            threshold: Threshold::AbsoluteCount {
-                weight: multisig_settings.threshold_absolute_count,
-            },
-            max_voting_period: MAX_MULTISIG_VOTING_PERIOD,
-        };
+        }
+        None => Err(ContractError::GuardianRequestNotFound {}),
+    }
+}
 
-        let instantiate_msg = WasmMsg::Instantiate {
-            admin: Some(env.contract.address.to_string()),
-            code_id: instantiation_code_id,
-            msg: to_binary(&multisig_instantiate_msg)?,
-            funds: multisig_settings.multisig_initial_funds,
-            label: "Wallet-Multisig".into(),
-        };
-        let msg = SubMsg::reply_always(instantiate_msg, MULTISIG_INSTANTIATE_ID);
+pub fn execute_request_update_guardians(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    request: Option<GuardiansUpdateMsg>,
+) -> Result<Response, ContractError> {
+    ensure_is_relayer_or_user(deps.as_ref(), &env, &info.sender)?;
 
-        Ok(Response::new()
-            .add_submessage(msg)
-            .add_attribute("action", "Updated wallet guardians: Multisig"))
-    } else {
-        MULTISIG_ADDRESS.save(deps.storage, &None)?;
-        Ok(Response::new().add_attribute("action", "Updated wallet guardians: Non-Multisig"))
+    if is_frozen(deps.as_ref())? {
+        return Err(ContractError::Frozen {});
     }
+
+    match request {
+        Some(r) => {
+            r.guardians
+                .verify_guardians(&deps.api.addr_humanize(&USER.load(deps.storage)?.addr)?)?;
+
+            PENDING_GUARDIAN_ROTATION.save(
+                deps.storage,
+                &GuardiansUpdateRequest::new(r.guardians, r.new_multisig_code_id, &env.block),
+            )?;
+        }
+        None => PENDING_GUARDIAN_ROTATION.remove(deps.storage),
+    }
+
+    Ok(Response::new().add_attribute("action", "Update guardians request created"))
```

Comment:
This attribute might be confusing if user is removing their previous request, especially if we are going to index events.